File: android/src/test/java/app/getbullet/marcus/RenderModel.kt
package app.getbullet.marcus

import android.graphics.Color
import android.text.Spanned
import android.text.style.AbsoluteSizeSpan
import android.text.style.BackgroundColorSpan
import android.text.style.ForegroundColorSpan
import android.text.style.LeadingMarginSpan
import app.getbullet.marcus.spans.MarkdownBackgroundSpan
import app.getbullet.marcus.spans.MarkdownBoldSpan
import app.getbullet.marcus.spans.MarkdownFontFamilySpan
import app.getbullet.marcus.spans.MarkdownGapSpan
import app.getbullet.marcus.spans.MarkdownItalicSpan
import app.getbullet.marcus.spans.MarkdownLineHeightSpan
import app.getbullet.marcus.spans.MarkdownStrikethroughSpan
import app.getbullet.marcus.spans.MarkdownUnderlineSpan

/**
 * Serializes the spans a formatted string carries into the shared render model.
 *
 * One line per span, `start..end  attribute`, sorted so the output does not
 * depend on the order the formatter happened to apply things in. The attribute
 * vocabulary is deliberately platform neutral -- `bold`, `color(#RRGGBB)`,
 * `indent(first,rest)` -- so an Android dump and an iOS dump of the same case
 * can be diffed against each other directly.
 *
 * Only public API is read. Making span fields visible purely so a test could
 * assert on them would be the test changing the thing it measures.
 */
object RenderModel {

  fun dump(text: CharSequence, spanned: Spanned): String {
    val lines = spanned.getSpans(0, spanned.length, Any::class.java)
      .mapNotNull { span ->
        val attribute = describe(span) ?: return@mapNotNull null
        val start = spanned.getSpanStart(span)
        val end = spanned.getSpanEnd(span)
        Triple(start, end, attribute)
      }
      .sortedWith(compareBy({ it.first }, { -it.second }, { it.third }))
      .map { (start, end, attribute) ->
        "%4d..%-4d %s".format(start, end, attribute)
      }

    if (lines.isEmpty()) {
      return "(no spans)"
    }

    return lines.joinToString("\n")
  }

  /** The shared attribute name for a span, or null for spans we do not model. */
  private fun describe(span: Any): String? = when (span) {
    is MarkdownBoldSpan -> "bold"
    is MarkdownItalicSpan -> "italic"
    is MarkdownStrikethroughSpan -> "strikethrough"
    is MarkdownUnderlineSpan -> "underline"
    is MarkdownFontFamilySpan -> "font-family"
    is MarkdownLineHeightSpan -> "line-height"
    // The rounded rectangle drawn behind a mention, not a plain background
    // colour -- iOS models the same thing as `MarkdownTextBackground`.
    is MarkdownBackgroundSpan -> "background-shape"
    // The space held open after a marker so the next container's gutter lands
    // between the two.
    is MarkdownGapSpan -> "gap"
    is ForegroundColorSpan -> "color(${hex(span.foregroundColor)})"
    is BackgroundColorSpan -> "background(${hex(span.backgroundColor)})"
    is AbsoluteSizeSpan -> "font-size(${span.size})"
    is LeadingMarginSpan ->
      "indent(${span.getLeadingMargin(true)},${span.getLeadingMargin(false)})"
    // Anything still unnamed is flagged rather than dropped: a span the model
    // has no word for is exactly what a reviewer needs to see.
    else -> "?${span.javaClass.simpleName.removePrefix("Markdown").removeSuffix("Span")}"
  }

  private fun hex(color: Int): String =
    "#%02X%02X%02X".format(Color.red(color), Color.green(color), Color.blue(color))
}

File: ios-tests/Tests/MarcusFormatterTests/RenderModel.swift
import UIKit

@testable import MarcusFormatter

/// Serializes an attributed string into the shared render model.
///
/// Deliberately the same vocabulary and line shape as
/// `android/src/test/.../RenderModel.kt` -- `bold`, `color(#RRGGBB)`,
/// `font-size(N)`, `indent(first,rest)`, `gap` -- so the two platforms'
/// baselines diff against each other, not only against themselves.
///
/// Getting there takes two adjustments, because the platforms expose formatting
/// through opposite shapes. `enumerateAttributes` yields *runs* -- maximal
/// stretches where the whole attribute dictionary is constant -- so a bold span
/// broken by a colour change arrives as three runs, where Android's `getSpans`
/// reports one span over the lot. Ranges are therefore coalesced per attribute
/// afterwards. And every run carries the inherited font, which Android never
/// reports because nothing added a span for it, so attributes equal to the
/// default are dropped.
enum RenderModel {

  static func dump(
    _ attributedString: NSAttributedString,
    defaults: [NSAttributedString.Key: Any] = [:]
  ) -> String {
    let defaultFont = defaults[.font] as? UIFont
    var byAttribute: [String: [NSRange]] = [:]
    let full = NSRange(location: 0, length: attributedString.length)

    attributedString.enumerateAttributes(in: full, options: []) { attributes, range, _ in
      for name in describe(attributes, defaultFont: defaultFont) {
        byAttribute[name, default: []].append(range)
      }
    }

    let lines = byAttribute
      .flatMap { name, ranges in coalesce(ranges).map { (name: name, range: $0) } }
      .sorted {
        ($0.range.location, -$0.range.length, $0.name)
          < ($1.range.location, -$1.range.length, $1.name)
      }
      .map { String(format: "%4d..%-4d %@", $0.range.location, $0.range.upperBound, $0.name) }

    return lines.isEmpty ? "(no spans)" : lines.joined(separator: "\n")
  }

  /// Merges ranges that touch, so an attribute interrupted only by a different
  /// attribute is reported once over its whole extent.
  private static func coalesce(_ ranges: [NSRange]) -> [NSRange] {
    let sorted = ranges.sorted { $0.location < $1.location }
    var merged: [NSRange] = []

    for range in sorted {
      if let last = merged.last, last.upperBound >= range.location {
        merged[merged.count - 1] = NSRange(
          location: last.location,
          length: max(last.upperBound, range.upperBound) - last.location
        )
      } else {
        merged.append(range)
      }
    }

    return merged
  }

  private static func describe(
    _ attributes: [NSAttributedString.Key: Any],
    defaultFont: UIFont?
  ) -> [String] {
    var described: [String] = []

    for (key, value) in attributes {
      switch key {
      case .font:
        guard let font = value as? UIFont else { break }
        if font.isBold { described.append("bold") }
        if font.isItalic { described.append("italic") }
        // Only what the formatter changed. The inherited size and family are on
        // every run and would drown the diff.
        if let defaultFont, font.pointSize != defaultFont.pointSize {
          described.append("font-size(\(Int(font.pointSize.rounded())))")
        }
        if let defaultFont, font.familyName != defaultFont.familyName {
          described.append("font-family(\(font.familyName))")
        }

      case .foregroundColor:
        if let color = value as? UIColor { described.append("color(\(hex(color)))") }

      case .backgroundColor:
        if let color = value as? UIColor { described.append("background(\(hex(color)))") }

      case .strikethroughStyle:
        described.append("strikethrough")

      case .underlineStyle:
        described.append("underline")

      case .marcusTextBackground:
        described.append("background-shape")

      // The box behind a whole code block, which Android models as the
      // `MarkdownCodeBlockSpan` it sets over the same range.
      case .marcusCodeBlock:
        described.append("code-block")

      // The circle drawn in place of an unordered item's marker, which Android
      // models as the `MarkdownBulletSpan` it sets over the same character.
      case .marcusListBullet:
        described.append("bullet")

      // Kerning is how the space after a marker is held open here; Android does
      // the same job with a MarkdownGapSpan. One name for one thing.
      case .kern:
        described.append("gap")

      case .paragraphStyle:
        guard let style = value as? NSParagraphStyle else { break }
        let first = Int(style.firstLineHeadIndent.rounded())
        let rest = Int(style.headIndent.rounded())
        if first != 0 || rest != 0 { described.append("indent(\(first),\(rest))") }

      case .marcusBlockquoteDepth:
        if let depth = value as? NSNumber { described.append("quote-depth(\(depth.intValue))") }

      case .marcusBlockquoteIndent:
        if let indent = value as? NSNumber {
          described.append("quote-indent(\(Int(indent.doubleValue.rounded())))")
        }

      // Set across the whole string as a marker for the `_textOf:equals:`
      // swizzle. Reporting it on every line would say nothing.
      case .marcusText:
        break

      default:
        described.append("?\(key.rawValue)")
      }
    }

    return described
  }

  private static func hex(_ color: UIColor) -> String {
    var red: CGFloat = 0
    var green: CGFloat = 0
    var blue: CGFloat = 0
    var alpha: CGFloat = 0
    color.getRed(&red, green: &green, blue: &blue, alpha: &alpha)

    return String(
      format: "#%02X%02X%02X",
      Int((red * 255).rounded()),
      Int((green * 255).rounded()),
      Int((blue * 255).rounded())
    )
  }
}

private extension UIFont {
  var isBold: Bool { fontDescriptor.symbolicTraits.contains(.traitBold) }
  var isItalic: Bool { fontDescriptor.symbolicTraits.contains(.traitItalic) }
}
